// stylelint-lsp/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommandTitles = exports.DisableRuleCommandIds = exports.CommandIds = void 0;
var CommandIds;
(function (CommandIds) {
    CommandIds["applyAutoFixes"] = "stylelint.applyAutoFixes";
})(CommandIds = exports.CommandIds || (exports.CommandIds = {}));
var DisableRuleCommandIds;
(function (DisableRuleCommandIds) {
    DisableRuleCommandIds["applyDisableRuleInline"] = "stylelint.applyDisableRuleInline";
    DisableRuleCommandIds["applyDisableRuleToFile"] = "stylelint.applyDisableRuleToFile";
    DisableRuleCommandIds["applyDisableRuleToLine"] = "stylelint.applyDisableRuleToLine";
    DisableRuleCommandIds["applyDisableRuleToRange"] = "stylelint.applyDisableRuleToRange";
})(DisableRuleCommandIds = exports.DisableRuleCommandIds || (exports.DisableRuleCommandIds = {}));
exports.CommandTitles = {
    [CommandIds.applyAutoFixes]: "Apply all stylelint fixes",
    [DisableRuleCommandIds.applyDisableRuleInline]: "Disable stylelint rule inline",
    [DisableRuleCommandIds.applyDisableRuleToLine]: "Disable stylelint rule on this line",
    [DisableRuleCommandIds.applyDisableRuleToFile]: "Disable stylelint rule for entire file",
    [DisableRuleCommandIds.applyDisableRuleToRange]: "Disable stylelint rule for selection",
};
